### PR TITLE
new method that only cache_if_truthy

### PR DESCRIPTION
**Pull request by drhenner**
### Summary

Caches the result of a method if the result is truthy

Hence if you do this the result will not be cached:
  cache_if_truthy("User:#{id}:falsy_method") { nil }

On the other hand this would be cached.
  cache_if_truthy("User:#{id}:truthy_method") { 'a string is truthy in ruby' }

Best used for things like `def has_user_visited_page?`

### CODE REVIEW

**Comment by rafaelfranca** [comment]
What is the alternative code if you didn't have this feature? And how frequent it is used in web applications?

This is not a pattern that I see being used with cache, usually you cache the value of something not if the value is true or false.

We avoid add new things to active support unless they are used in the framework or by a large number of applications, so we need to know why do you think this feature is useful to take any decision.

**Comment by rafaelfranca** [comment]
Ah, forgot to say, thank you for the pull request 😄

**Comment by drhenner** [comment]
Sorry I wasn't expecting a question so quickly.

I have now had to add this functionality to my last two apps.  One of my specific use cases was if a user got a subscription I'd cache the result from stripe.com for the next several hours.  Otherwise checking to see if every user within an account had a subscription would be very expensive.

File: activesupport/lib/active_support/cache.rb [hunk 1]
```diff
@@ -422,6 +422,34 @@ def exist?(name, options = nil)
...
+      #
+      def cache_if_truthy(key, options = nil)
+        if block_given?
+          result = read(key, options)
+          if !result || (options && options[:force])
+            result = yield
+            write(key, result, options) if result || (options && options[:force])
+          end
+        elsif options && options[:force]
+          raise ArgumentError, 'Missing block: Calling `Cache#cache_if_true` with `force: true` requires a block.'
```
Comment (bryanaka) [inline activesupport/lib/active_support/cache.rb]: Did you mean `Cache#cache_if_truthy` here?

Comment (drhenner) [inline activesupport/lib/active_support/cache.rb]: oops, yup

**Comment by arthurnn** [comment]
I am not too sure about adding a new API like that to AS cache, I, personally, never had such use case.
@dhh @jeremy have you had to implement this type of thing a lot before?

My general rule of thumb is, if this is not something that most of the rails apps will use, and it is easy enough to implement it yourself in your app, i would say this should not go to AS.

**Comment by dhh** [comment]
I think this is a fine candidate for a gem. Never had this need before, and generally dislike the if/unless versions where we have them. I think we have some in HTML tag generation.

**Pull request closed without merging**